### PR TITLE
Chapter1/homework/5

### PR DESCRIPTION
**Pull request by jeanroyce**
This one was hard to get pass in all 42 test, but I did it.

One question, why this code wont work properly?

```C#
private static void Revert(int[] array)
{
	var tempArray = array;

	for (int i = 0; i < array.Length; i++)
	{
		array[i] = tempArray[^(i + 1)];
	}
}
```
Anotação 2020-04-12 233128

### CODE REVIEW

**Comment by Almantask** [comment]
> This one was hard to get pass in all 42 test, but I did it.
> 
> One question, why this code wont work properly?
> 
> ```cs
> private static void Revert(int[] array)
> {
> 	var tempArray = array;
> 
> 	for (int i = 0; i < array.Length; i++)
> 	{
> 		array[i] = tempArray[^(i + 1)];
> 	}
> }
> ```
> 
> Anotação 2020-04-12 233128

It's not working because arrays **are passed by reference**. tempArray and array are still pointing to the same reference (the same array). In order to create a new reference, you need to clone the array or create a new array and copy all the values.

What actually happened is:
iteration 1:
``{2 0 1}`` - places 2 at the start of array because it's the last element.
iteration 2:
``{ 2 0 1}`` - places 0 at the 2nd position of array (so nothing happens) because it's second last element
iteration 3:
``{2 0 2}`` - places 2 at the end of array because it's the first element (based on the already modified array)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 1]
```diff
@@ -9,7 +9,25 @@ public static class ArrayOperations
         /// <param name="array">Input array in a random order.</param>
         public static void Sort(int[] array)
         {
-            // ToDo: implement.
+            if (IsNullOrEmpty(array)) return;
+
+            bool isSorted;
+            do
+            {
+                isSorted = true;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs:17]: It's the opposite I think. Is it sorted at the start? No.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 2]
```diff
@@ -9,7 +9,25 @@ public static class ArrayOperations
         /// <param name="array">Input array in a random order.</param>
         public static void Sort(int[] array)
         {
-            // ToDo: implement.
+            if (IsNullOrEmpty(array)) return;
+
+            bool isSorted;
+            do
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs:15]: Use 2 for loops instead.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 3]
```diff
@@ -9,7 +9,25 @@ public static class ArrayOperations
...
+            {
+                isSorted = true;
+                for (int i = 0; i < array.Length - 1; i++)
+                {
+                    if (array[i] > array[i + 1])
+                    {
+                        int oldIndexValue = array[i + 1];
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs:22]: How many cups do you need to move wine from cup 1 to cup 2?

Comment (jeanroyce) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: 3 is enough

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 1 above)
Comment (jeanroyce) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: I should have done the opposite.
I was about to lost my mind in this homework and I didn't pay attention, but now I refactored this part of code.

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 2 above)
Comment (jeanroyce) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: Done now with 2 for loops.

**Comment by jeanroyce** [comment]
> It's not working because arrays **are passed by reference**. tempArray and array are still pointing to the same reference (the same array). In order to create a new reference, you need to clone the array or create a new array and copy all the values.

I get it, thanks!

**Comment by Almantask** [comment]
Looks good

**Pull request closed without merging**